### PR TITLE
Add microbatch context helper

### LINKED ISSUE #5616: MFSDP v2 should support HSDP and HFSDP placements
MFSDP v2 currently rejects the placement differences needed for HSDP/HFSDP.

`megatron_fsdp/experimental/parameter_group.py` requires `main_grad` and `main_weight` to have identical placements and raises until HSDP/HFSDP support is implemented. That blocks hybrid sharding modes where weights, gradients, and optimizer state intentionally use different inner-DP / outer-DP layouts.

Implementation detail:

- First microbatch: all-gather across DP-outer.
- Last microbatch: reduce-scatter or all-reduce across DP-outer.
- DP-outer communication should probably run on a separate stream so it can overlap with DP-inner communication.

Expected outcome:

- Define MFSDP v2 layout semantics for HSDP and HFSDP.
- Support HSDP outer-DP replication / gradient synchronization.
- Support HFSDP optimizer-state sharding across inner and outer DP ranks.
- Add focused tests for the required placement and collective transitions.

Development PRs:
- #5652

### PR DESCRIPTION
**Pull request by wujingyue**
## Summary
- Add `FsdpContext` with `is_last_microbatch` state for experimental FSDP.
- Add `microbatch(module, is_last)` to scope microbatch state across FSDP contexts.
- Add `FsdpModule.context` and root ownership tracking for lazily initialized context state.
- Expose only `fully_shard` and `microbatch` from the experimental package root.
- Add coverage for shared child/root contexts and unwrapped-parent child contexts.

## Testing
- `PATH=".venv/bin:$PATH" BASE_REF=main CHECK_ONLY=true SKIP_DOCS=false bash tools/autoformat.sh`
- `git diff --check`
- `rg -n "is_first_microbatch|is_first\\b" megatron/core/distributed/fsdp/src/megatron_fsdp/experimental tests/unit_tests/distributed/megatron_fsdp/test_context.py tests/unit_tests/distributed/megatron_fsdp/test_experimental_fully_shard.py` (no matches)
- `/opt/venv/bin/python -m torch.distributed.run --nproc-per-node 2 -m pytest -q tests/unit_tests/distributed/megatron_fsdp/test_context.py tests/unit_tests/distributed/megatron_fsdp/test_experimental_fully_shard.py --tb=short --disable-warnings -rN` (`13 passed`)

### CODE REVIEW

File: megatron/core/distributed/fsdp/src/megatron_fsdp/experimental/fully_shard.py [hunk 1]
```diff
@@ -63,9 +66,43 @@ def fully_shard(
...
 
 
+@contextmanager
+def microbatch(module: nn.Module, is_first: bool) -> Iterator[None]:
+    """Scope experimental FSDP state to one microbatch.
+
+    Args:
+        module: Module tree whose experimental FSDP roots should use this microbatch state.
+        is_first: Whether forwards in this scope are for the first microbatch.
+    """
```
Comment (shjwudp) [inline megatron/core/distributed/fsdp/src/megatron_fsdp/experimental/fully_shard.py:76]: Could we also add a set_is_last_backward API to match the corresponding PyTorch FSDP2 API?
https://docs.pytorch.org/docs/2.12/distributed.fsdp.fully_shard.html#torch.distributed.fsdp.FSDPModule.set_is_last_backward

Comment (wujingyue) [reply megatron/core/distributed/fsdp/src/megatron_fsdp/experimental/fully_shard.py:76]: I prefer to add it **if/when** we need it. https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it

File: megatron/core/distributed/fsdp/src/megatron_fsdp/experimental/module.py [hunk 2]
```diff
@@ -26,10 +26,17 @@
 from .placement import MeshAxis, Placements
 
 
+class FsdpContext:
+    """Runtime state shared by one experimental FSDP subtree."""
+
+    is_first_microbatch: bool = True
```
Comment (wujingyue) [inline megatron/core/distributed/fsdp/src/megatron_fsdp/experimental/module.py]: From @Autumn1998: Add is_last_microbatch here as well for HSDP

**Review by shjwudp — APPROVED** [review]

File: megatron/core/distributed/fsdp/src/megatron_fsdp/experimental/module.py (hunk 2 above)
Comment (wujingyue) [reply megatron/core/distributed/fsdp/src/megatron_fsdp/experimental/module.py]: Update: I'll replace this with is_last_microbatch

File: megatron/core/distributed/fsdp/src/megatron_fsdp/experimental/__init__.py [hunk 3]
```diff
@@ -15,19 +15,16 @@
 """Experimental Megatron-FSDP implementation."""
 
 from .dbuffer import DBuffer
-from .fully_shard import fully_shard
-from .module import FsdpModule
-from .parameter_group import FsdpParameterGroup
```
Comment (wujingyue) [inline megatron/core/distributed/fsdp/src/megatron_fsdp/experimental/__init__.py:20]: Removed because they are not needed

**Review by jaredcasper — APPROVED** [review]

**Review by ericharper — APPROVED** [review]
